File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys
input = sys.stdin.readline

N = int(input())

left = [0] * (N+1)
right = [0] * (N+1)
tree = {}
for i in range(1, N+1):

    x, y, z = map(str, input().split())
    tree[x] = [y, z]

preorder = []
inorder = []
postorder = []

def preorder_travers(T):
    if T != '.':
        preorder.append(T)
        preorder_travers(tree[T][0])
        preorder_travers(tree[T][1])
    return preorder

print("".join(preorder_travers('A')))

def inorder_travers(T):
    if T != '.':
        inorder_travers(tree[T][0])
        inorder.append(T)
        inorder_travers(tree[T][1])
    return inorder

print("".join(inorder_travers('A')))

def postorder_travers(T):
    if T != '.':
        postorder_travers(tree[T][0])
        postorder_travers(tree[T][1])
        postorder.append(T)
    return postorder

print("".join(postorder_travers('A')))

File: 백준/Silver/1991. 트리 순회/트리 순회.js
let fs = require("fs");
let input = fs.readFileSync("/dev/stdin");
// let input = fs.readFileSync("./input.txt");
input = input.toString().trim().split("\n");

const N = parseInt(input[0]);
arr = Array.from({ length: N + 1 }, () => []);
for (let i = 1; i <= N; i++) {
  const [P, L, R] = input[i].split(" ").map((x) => x.charCodeAt() - 64);

  if (L > 0) {
    arr[P].push(L);
  } else {
    arr[P].push(".");
  }

  if (R > 0) {
    arr[P].push(R);
  } else {
    arr[P].push(".");
  }
}

preVal = "";
function Preorder(x) {
  if (x !== ".") {
    let point = x + 64;
    preVal += String.fromCodePoint(point);
    Preorder(arr[x][0]);
    Preorder(arr[x][1]);
  }
}

Preorder(1);
console.log(preVal);

inVal = "";
function Inorder(x) {
  if (x !== ".") {
    Inorder(arr[x][0]);
    let point = x + 64;
    inVal += String.fromCodePoint(point);
    Inorder(arr[x][1]);
  }
}

Inorder(1);
console.log(inVal);

postVal = "";
function Postorder(x) {
  if (x !== ".") {
    Postorder(arr[x][0]);
    Postorder(arr[x][1]);
    let point = x + 64;
    postVal += String.fromCodePoint(point);
  }
}

Postorder(1);
console.log(postVal);
